parser/test/test_parse_authenticate_body.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2020 Maksym Sobolyev
+ *
+ * This file is part of opensips, a free SIP server.
+ *
+ * opensips is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version
+ *
+ * opensips is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,USA
+ */
+
+#include <tap.h>
+
+#include "../../str.h"
+#include "../../ut.h"
+
+#include "../../parser/parse_authenticate.h"
+
+#include "test_parse_authenticate_body.h"
+#include "test_oob.h"
+
+void test_parse_authenticate_body_oob(const str *, enum oob_position where, void *);
+
+void test_parse_authenticate_body(void)
+{
+	struct authenticate_body auth;
+	int i;
+	struct tts {
+		const str ts;
+		int tres;
+		int aflags;
+		const char *anonce;
+		const char *aopaque;
+		const char *arealm;
+	} tset[] = {
+		{.ts = str_init("Digest realm=\"[::1]\",nonce=\"ak/bKmGcoPPWdj0AUWv/ldViLInmkiJ2Kct5p/LapNo\","
+		    "qop=auth,algorithm=SHA-512-256"), .tres = -1},
+		{.ts = str_init("Digest realm=\"[::1]\",nonce=\"kp5XbciCMxVbeZm2d58YZCfaAjW/2T7XtuYwIeZoz1o\","
+		    "qop=auth,algorithm=SHA-256"), .tres = -1},
+		{.ts = str_init("Digest stale=false,realm=\"[::1]\",nonce=\"esWk1wFa4bUBKzkmfKId++Y83eWzD9edBCGTwLV4Juk\","
+		    "qop=auth,algorithm=MD5"), .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5,
+		    .anonce = "esWk1wFa4bUBKzkmfKId++Y83eWzD9edBCGTwLV4Juk", .arealm = "[::1]"},
+		{.ts = str_init("Digest realm=\"sip.test.com\",qop=\"auth\",opaque=\"1234567890abcedef\","
+		    "nonce=\"145f5ca9aac6f0b9f93433188d446ae0d9f91a6ff80\",algorithm=MD5,stale=true"),
+		    .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5 | AUTHENTICATE_STALE,
+		    .anonce = "145f5ca9aac6f0b9f93433188d446ae0d9f91a6ff80", .aopaque = "1234567890abcedef",
+		    .arealm = "sip.test.com"},
+		{.ts = str_init("DiGeSt\r\n\trealm=\"a\",\r\n\tqop=\"auth-int, auth\",\r\n\tnonce=\"n\",\r\n\topaque=\"0\",\r\n\t"
+		    "algoriTHm=md5"), .tres = 0, .aflags = QOP_AUTH | AUTHENTICATE_MD5 | QOP_AUTH_INT,
+		    .anonce = "n", .aopaque = "0", .arealm = "a"},
```

Comment:
All is good here -- but here is some personal feedback (nitpicking):

* I found it a bit hard to tell why each `-1` test was expected to fail.  Eventually, I understood that it was due to the new RFC 8780 algorithms
* the entire test set looks like a wall of text and is hard to read.  Maybe adding an empty-line and/or comments between each test would mitigate this problem?